File: packages/detector-core-ts/src/detectors/authorship.ts
/**
 * Authorship-artifact heuristics: the "too clean / too polished" tells.
 *
 * Mirrors the Python AuthorshipArtifactsDetector. Targets natural AI code that
 * lacks tutorial-style tells, via two empirically-derived signals:
 *   - absence of human maintenance artifacts (TODO/FIXME/noqa/commented code)
 *   - unusually polished, full-sentence comments
 *
 * These are weak, probabilistic signals (a meticulous human looks AI-like here)
 * and natural AI can evade them entirely — see docs/limitations.md.
 */

import type { Signal } from "../models.js";
import type { Scan } from "../ingestion/models.js";
import { scanAnalyzable } from "../ingestion/models.js";
import { clamp, mean } from "../utils/text.js";
import { stripComments } from "../parsing/languages.js";
import { fixed } from "../utils/format.js";
import { Detector, unitLoc, type AnalysisUnit } from "./base.js";

const ARTIFACT_RE = /\b(TODO|FIXME|HACK|XXX)\b/g;
const PRAGMA_RE = /(noqa|type:\s*ignore|pylint:|mypy:|pragma|# *type:)/g;
const WELLFORMED_RE = /^[A-Z][^\n]{6,}[.!?]$/;
const CODEISH_RE = /(=|\(|\)|\bdef\b|\bclass\b|\bimport\b|\breturn\b|\bself\.|\[|\])/;

function cleanComments(source: string, language: string): string[] {
  const [, comments] = stripComments(source, language);
  return comments.map((c) => c.trim()).filter((c) => c.length > 0);
}

function isCommentedCode(text: string): boolean {
  // No AST in the browser; a conservative structural heuristic (kept close to
  // the Python ast-based check so cross-engine parity holds on clean corpora).
  if (!CODEISH_RE.test(text)) return false;
  return /[;{}]\s*$/.test(text) || /^(def |class |import |from |return |if |for |while |[\w.]+\s*=)/.test(text);
}

function count(re: RegExp, s: string): number {
  const m = s.match(re);
  return m ? m.length : 0;
}

export class AuthorshipArtifactsDetector extends Detector {
  override name = "authorship_artifacts";
  override description = "Absence of human maintenance artifacts and unusually polished comments.";

  override unitSignals(unit: AnalysisUnit): Signal[] {
    if (unit.is_documentation || unitLoc(unit) < 12) return [];
    const comments = cleanComments(unit.source, unit.language);
    if (comments.length < 4) return [];
    const wf = mean(comments.map((c) => (WELLFORMED_RE.test(c) ? 1 : 0)));
    const score = clamp(0.5 + 1.0 * (wf - 0.22));
    const ev = [];
    if (score > 0.62) {
      const sample = comments.find((c) => WELLFORMED_RE.test(c));
      ev.push(
        this.evidence(
          "comment_polish",
          `${Math.round(wf * 100)}% of comments are polished full sentences (humans tend ` +
            `to write terse, fragmentary comments).`,
          { severity: score, path: unit.path, start_line: unit.start_line, snippet: sample },
        ),
      );
    }
    return [
      this.signal("comment_polish", score, {
        confidence: clamp(comments.length / 8),
        weight: 1.3,
        reason: `${Math.round(wf * 100)}% of comments are complete, well-formed sentences.`,
        evidence: ev,
      }),
    ];
  }

  override repoSignals(scan: Scan, _units: AnalysisUnit[]): Signal[] {
    const files = scanAnalyzable(scan).filter((f) => !f.is_documentation);
    const totalLoc = files.reduce((a, f) => a + f.source.split("\n").length, 0);
    if (totalLoc < 40) return [];
    const full = files.map((f) => f.source).join("\n");
    const kloc = Math.max(0.3, totalLoc / 1000);
    const artifacts = (count(ARTIFACT_RE, full) + count(PRAGMA_RE, full)) / kloc;

    const score = clamp(0.585 - 0.5 * Math.min(artifacts, 1.0));
    const ev = [];
    if (artifacts < 0.05 && totalLoc > 150) {
      ev.push(
        this.evidence(
          "missing_human_artifacts",
          `No human maintenance artifacts (TODO/FIXME/noqa/commented-out code) found across ` +
            `${totalLoc} lines; human codebases of this size usually contain some.`,
          { severity: score },
        ),
      );
    }
    const signals: Signal[] = [
      this.signal("missing_human_artifacts", score, {
        confidence: clamp(totalLoc / 600),
        weight: 1.0,
        reason: `${fixed(artifacts, 2)} maintenance artifacts per KLOC (absence leans AI; presence leans human).`,
        evidence: ev,
      }),
    ];

    const lang = scan.files.some((f) => f.language === "python") ? "python" : "javascript";
    const comments = cleanComments(full, lang);
    if (comments.length >= 8) {
      const cc = mean(comments.map((c) => (isCommentedCode(c) ? 1 : 0)));
      signals.push(
        this.signal("no_commented_out_code", clamp(0.54 - 1.3 * cc), {
          confidence: clamp(comments.length / 25),
          weight: 0.5,
          reason: `${Math.round(cc * 100)}% of comments are commented-out code (absence leans AI).`,
        }),
      );
    }
    return signals;
  }
}

File: packages/detector-core/aiprojectdetector/detectors/authorship.py
"""Authorship-artifact heuristics: the 'too clean / too polished' tells.

Unlike the LLM-fingerprint detector (which keys on tutorial-style narration),
this detector targets *natural* AI code that lacks obvious tells. It is based on
an empirical comparison of real human modules (CPython stdlib + Flask) against
independent AI re-implementations of the same functionality:

  * Humans leave **maintenance artifacts** — TODO/FIXME/HACK, ``# noqa``,
    ``# type: ignore``, commented-out experiments. Natural AI output had
    essentially none.
  * Humans write **terse, fragmentary comments**; AI writes grammatically
    complete, capitalized, full-sentence comments. (Measured: ~0.15 well-formed
    comment ratio for humans vs ~0.50 for AI.)

These are still weak, probabilistic signals — a meticulous human who leaves no
TODOs and writes polished comments will look more AI-like here. That trade-off
is intentional and documented (see docs/limitations.md): catching "clean" AI
necessarily raises the false-positive rate on very tidy human code.
"""

from __future__ import annotations

import ast
import re

from ..ingestion.models import Scan
from ..models import Signal
from ..parsing.languages import strip_comments
from ..utils.text import clamp, mean
from .base import AnalysisUnit, Detector

_ARTIFACT_RE = re.compile(r"\b(TODO|FIXME|HACK|XXX)\b")
_PRAGMA_RE = re.compile(r"(noqa|type:\s*ignore|pylint:|mypy:|pragma|# *type:)")
_WELLFORMED_RE = re.compile(r"^[A-Z][^\n]{6,}[.!?]$")
_CODEISH_RE = re.compile(r"(=|\(|\)|\bdef\b|\bclass\b|\bimport\b|\breturn\b|\bself\.|\[|\])")


def _clean_comments(source: str, language: str) -> list[str]:
    _, comments = strip_comments(source, language)
    return [c.strip() for c in comments if c.strip()]


def _is_commented_code(text: str, language: str) -> bool:
    """Detect commented-out code (a human tell). Python uses ast; others regex."""
    if not _CODEISH_RE.search(text):
        return False
    if language == "python":
        for candidate in (text, text + " pass" if text.rstrip().endswith(":") else text):
            try:
                ast.parse(candidate)
                return True
            except SyntaxError:
                continue
        return False
    # Generic: a line ending in ; or { or } that also looks code-ish.
    return bool(re.search(r"[;{}]\s*$", text))


class AuthorshipArtifactsDetector(Detector):
    name = "authorship_artifacts"
    description = "Absence of human maintenance artifacts and unusually polished comments."

    def unit_signals(self, unit: AnalysisUnit) -> list[Signal]:
        if unit.is_documentation or unit.loc < 12:
            return []
        comments = _clean_comments(unit.source, unit.language)
        if len(comments) < 4:
            return []
        wf = mean([1.0 if _WELLFORMED_RE.match(c) else 0.0 for c in comments])
        # Humans ~0.15, AI ~0.5+. Map two-sided around ~0.22.
        score = clamp(0.5 + 1.0 * (wf - 0.22))
        ev = []
        if score > 0.62:
            sample = next((c for c in comments if _WELLFORMED_RE.match(c)), None)
            ev.append(
                self.evidence(
                    "comment_polish",
                    f"{wf:.0%} of comments are polished full sentences (humans tend to "
                    f"write terse, fragmentary comments).",
                    severity=score,
                    path=unit.path,
                    start_line=unit.start_line,
                    snippet=sample,
                )
            )
        return [
            self.signal(
                "comment_polish",
                score,
                confidence=clamp(len(comments) / 8),
                weight=1.3,
                reason=f"{wf:.0%} of comments are complete, well-formed sentences.",
                evidence=ev,
            )
        ]

    def repo_signals(self, scan: Scan, units: list[AnalysisUnit]) -> list[Signal]:
        files = [f for f in scan.analyzable() if not f.is_documentation]
        total_loc = sum(f.loc for f in files)
        if total_loc < 40:
            return []
        full = "\n".join(f.source for f in files)
        kloc = max(0.3, total_loc / 1000)
        artifacts = (len(_ARTIFACT_RE.findall(full)) + len(_PRAGMA_RE.findall(full))) / kloc

        # Absence of any maintenance artifacts leans AI; presence leans human.
        score = clamp(0.585 - 0.5 * min(artifacts, 1.0))
        conf = clamp(total_loc / 600)
        ev = []
        if artifacts < 0.05 and total_loc > 150:
            ev.append(
                self.evidence(
                    "missing_human_artifacts",
                    f"No human maintenance artifacts (TODO/FIXME/noqa/commented-out code) "
                    f"found across {total_loc} lines; human codebases of this size usually "
                    f"contain some.",
                    severity=score,
                )
            )
        signals = [
            self.signal(
                "missing_human_artifacts",
                score,
                confidence=conf,
                weight=1.0,
                reason=f"{artifacts:.2f} maintenance artifacts per KLOC "
                f"(absence leans AI; presence leans human).",
                evidence=ev,
            )
        ]

        # Commented-out code is a (weak) human tell; total absence in a sizable
        # codebase nudges toward AI.
        comments = _clean_comments(full, "python" if scan.languages().get("python") else "javascript")
        if len(comments) >= 8:
            cc = mean([1.0 if _is_commented_code(c, "python") else 0.0 for c in comments])
            cscore = clamp(0.54 - 1.3 * cc)
            signals.append(
                self.signal(
                    "no_commented_out_code",
                    cscore,
                    confidence=clamp(len(comments) / 25),
                    weight=0.5,
                    reason=f"{cc:.0%} of comments are commented-out code "
                    f"(humans leave some; absence leans AI).",
                )
            )
        return signals
